File: _eunjin/6236_용돈_관리.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
arr = list(int(input()) for _ in range(N))


# 이분탐색?
left = max(arr)
right = sum(arr)  # 한번에 전체 금액 뽑는 경우

while left <= right:
    mid = (left + right) // 2

    out = mid  # 인출 금액
    cnt = 1  # 인출 횟수

    for n in arr:
        if n > out:  # 다시 인출
            cnt += 1
            out = mid
        out -= n  # 사용

    if cnt > M:
        left = mid + 1
    else:
        right = mid - 1

print(mid)

File: _hyolim/6236_용돈관리.cpp
#include <iostream>

using namespace std;

int arr[100001]={0,};
int n,m;

void input(){
	cin>>n>>m;
	for(int i=0;i<n;i++){
		cin>>arr[i];
	}
}

bool checkIsAvailable(long num){
	long cnt=0;
	long money=0;
	for(int i=0;i<n;i++){
		// cout<<arr[i]<<" "<<money<<" "<<cnt<<"\n" ;

		if(arr[i]>num) return false;
		if(money-arr[i]<0){
			cnt++;
			money=num;
		}
		money-=arr[i];

	}
	if(money<0) return false;
	if(cnt>m) return false;
	return true;
}

void solve(){
	// 이분탐색
	long l=0;
	long r=987654321987654321;
	long mid=(l+r)/2;

	while(l<=r){
		// cout<<l<<" "<<mid<<" "<<r<<"\n";
		if(checkIsAvailable(mid)){
			// cout<<"*";
			r=mid-1;
		}else{
			l=mid+1;
		}
		mid=(l+r)/2;
	}
	cout<<l;
}

int main(){
	input();
	solve();
}